Add Rectangular Area Light Source

CookieBadger opened the pull request:
This PR adds a rectangular area light source, implemented via Linearly Transformed Cosines (LTC) with textures. It makes use of a Lookup-Table added in the form of two .dds files, and adds some fields to the `LightData` struct in the light storage, and adds a new atlas for area light textures.

- Closes https://github.com/godotengine/godot-proposals/issues/11450.
- Closes https://github.com/godotengine/godot-proposals/issues/13214

<img width="2047" height="1534" alt="image" src="https://github.com/user-attachments/assets/4de42d95-af99-4949-8fd4-889dd50908cc" />

*Image above made by: @passivestar*

![LTCHeitz_compressed](https://github.com/user-attachments/assets/70b67daa-6eea-4e44-80c9-8767cea5f657)
<img width="2150" height="1218" alt="Screenshot 2026-03-15 004153" src="https://github.com/user-attachments/assets/515501b6-be90-4ac3-b702-bab119e428c9" />
<img width="1237" height="693" alt="black_hole" src="https://github.com/user-attachments/assets/e36c0dd0-3c46-4fdb-b6e5-cfb786617765" />
<img width="1572" height="1102" alt="image" src="https://github.com/user-attachments/assets/627fa2a5-7812-4478-bff7-62dd00dbd0d8" />
<img width="1024" height="890" alt="image" src="https://github.com/user-attachments/assets/9963ed1a-7126-4e16-852a-16098156dffa" />

- [x] Integrate lights in VolumetricFog
- [x] Integrate lights in LightMaps
- [x] Integrate lights in VoxelGI
- [x] Integrate lights in SDFGI
- [x] Integrate material effects (Clearcoat, Toon Diffuse & Specular, Rim, Backlight, Transmittance/SSS)
- [x] Implement texturing the light quad

Here's a list of steps on how to improve the feature in the future:
- Support importing area lights from blender files
- Integrate anisotropy (https://github.com/AakashKT/LTC-Anisotropic/tree/main)
- Integrate in Vertex Shading (if there's a demand)

Here's how shadows can be improved:
- Add shadows in Compatibility renderer (requires adding paraboloid or alternative shadow mapping to the GLES3 shader).
- Improve accuracy of paraboloid rendering in Forward renderers (currently area lights use the optimized version that was implemented for omni lights, but a more accurate represenation might better prevent leaks)
- Add monte carlo shadows in lightmaps - it would need some parameter for how many rays should be traced. This would be a way to actually get realistic shadows with area lights, so maybe something we would want.

One known issue is that specular light can leak through to a face that faces away from the area when viewing at a steep angle and a low roughness value. This is a literal edge case, and fixing it would likely require additional horizon clipping, so maybe we can live with the leak for now. It is present in the source LTC implementation, and the implementation in Unity, and I didn't notice until very recently.

The main sources for this technique are https://eheitzresearch.wordpress.com/415-2/ and https://blog.selfshadow.com/publications/s2016-advances/
In theory, it can be extended to arbitrary polygons, disks, and lines, to serve as the light surface (https://eheitzresearch.wordpress.com/757-2/), but I would worry about adding those in the future.  

In the process of making this PR I investigated different techniques for lighting and shading. I investigated a Monte-Carlo sampling approach, and a most-representative point sampling approach, but both were outperformed by LTC, and contemporary research of non-raytraced dynamic area lights still seems to also focus around LTC.
I also investigated shadow sampling with many shadow maps on a new atlas, and adaptive sampling techniques, but those proved to be unusable due to the added implementation complexity. I figured the most suitable shadow technique was using 4 shadow maps per light and directional sampling with PCF to obtain a smooth result, but the shape of the penumbra was strangely warped, and the performance was not good. If you are curious, you may check it out at https://github.com/CookieBadger/godot/tree/area-light-quadriscopic-shadows. Another way to shadow area lights would of course be ray tracing (https://eheitzresearch.wordpress.com/705-2/), but I'm not aware of any developments of the engine in that direction. So, I decided to just integrate PCSS shadows for area lights, which are not physically accurate, but can look decent.

You can find a writeup and builds with several example scenes at https://github.com/CookieBadger/master-thesis-artifacts, but those already outdated as they are based on Godot 4.3. You'll also find a Thesis.pdf there, which is my Master's Thesis about this topic.

Any help with improving and maintaing this feature will be greatly appreciated.

Calinou commented:
> Also, CI seems to fail due to some GDExtension stuff that I am unfamiliar with, I'd appreciate a pointer to the fix.

The failures are because you reordered some enums around (specifically the Light3D `PARAM_*` enum), which breaks compatibility with existing projects. Make sure you only add new values at the end; never move existing values or remove them.

Calinou reviewed (commented):
Tested locally, it works as expected. This is a great start :slightly_smiling_face: 

**Testing project:** [test_pr_108219.zip](https://github.com/user-attachments/files/21040506/test_pr_108219.zip)

Some feedback:

- Area width and height should use a single `area` Vector2 property instead of two separate float properties. While doing so, make sure to use `PROPERTY_HINT_LINK` for the property so you can adjust the width proportionally to the current height (and vice versa). You can clamp the value to positive numbers in the `set_area()` setter (this ensures it also affects script-based usage).
- Gizmo behavior and icons look good to me.
- I've only tested this in Forward+, as Mobile will print a lot of errors and not render anything. On Compatibility, area lights are not supported but rendering works correctly otherwise.
  - There should be a node configuration warning on AreaLight3D when in Compatibility, similar to the Decal node's existing node configuration warning.
- Like with lights that use a dual paraboloid shadow mode, meshes need to be subdivided enough (relative to the size of the area light) for shadows to look correct:

https://github.com/user-attachments/assets/eb14853e-f00f-4825-a93d-ac91255f6c8f

This is most noticeable on shadows with a low blur value:

https://github.com/user-attachments/assets/e5e96398-6670-4a4c-b902-cad1f45dfe6a

- Light/shadow distance fade works correctly.
- Volumetric fog doesn't take area lights into account yet.
- **Hard** positional shadow quality looks the same as **Soft Very Low**. Ideally, **Hard** should disable shadow blurring entirely like it does with other light types to maximize performance on low-end GPUs.
- **High** and **Ultra** soft shadow qualities can be very expensive with lights that have a high blur value (even `1.0` is quite blurry out of the box), so beware. Consider using TAA or FSR2 to get temporal smoothing (and keep shadow filter quality at its default **Low**), which will give you smoother shadows without an increased performance cost.
- Shadow opacity will behave differently to other light types, as the shadow won't fade in an uniform manner across the whole light. This may be desired behavior, but it might be worth documenting to avoid surprises:

https://github.com/user-attachments/assets/cc4cd507-84ef-491c-b1d5-210907fbfc21

- When you move some shadow casters, they can sometimes be in a position where they will self-shadow. This is most noticeable at low blur values (below `0.3` in the test project). Reloading the scene or restarting the project does *not* fix the self-shadowing, so this does not appear to be an issue related to shadow updates sometimes not occurring when they should.

https://github.com/user-attachments/assets/131dee5c-c6c1-40d0-bcfc-2d8e68a7dbc

The issue will occur even if you're moving another shadow caster:

https://github.com/user-attachments/assets/2a254c09-f2ad-47b6-a9ee-9c10f159369e

- Specular contribution is set to `0.5` by default, which leads to area lights not having specular contributions as strong as omni and spot lights:

![image](https://github.com/user-attachments/assets/8691297e-fd4c-41fa-bdfe-357a1dd5a9cf)

It's more in line with other light types when set to `1.0`, so I suggest changing the default to that:

![image](https://github.com/user-attachments/assets/ea361632-ba79-43d9-8a62-61ff05d94eb8)

Calinou commented on doc/classes/AreaLight3D.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="AreaLight3D" inherits="Light3D" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="area_attenuation" type="float" setter="set_param" getter="get_param" default="1.0">
+		</member>
+		<member name="area_height" type="float" setter="set_param" getter="get_param" default="1.0">
+		</member>
+		<member name="area_range" type="float" setter="set_param" getter="get_param" default="5.0">
+		</member>
+		<member name="area_width" type="float" setter="set_param" getter="get_param" default="1.0">
+		</member>
+		<member name="light_size" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.5" />
+		<member name="light_specular" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.5" />
+		<member name="shadow_bias" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.03" />
+		<member name="shadow_normal_bias" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="1.0" />
+	</members>
```
```suggestion
	<brief_description>
		An area light, such as a neon light tube or a screen.
	</brief_description>
	<description>
		An area light is a type of [Light3D] node that emits light over a 2-dimensional area, in the shape of a rectangle. The light is attenuated through the distance. This attenuation can be configured by changing the energy, [member area_attenuation], [member area_range]
	</description>
	<tutorials>
	</tutorials>
	<members>
		<member name="area_attenuation" type="float" setter="set_param" getter="get_param" default="1.0">
			Controls the distance attenuation function for area lights.
			A value of [code]0.0[/code] will maintain a constant brightness through most of the range, but smoothly attenuate the light at the edge of the range. Use a value of [code]2.0[/code] for physically accurate lights as it results in the proper inverse square attenuation.
			[b]Note:[/b] Setting attenuation to [code]2.0[/code] or higher may result in distant objects receiving minimal light, even within range. For example, with a range of [code]4096[/code], an object at [code]100[/code] units is attenuated by a factor of [code]0.0001[/code]. With a default brightness of [code]1[/code], the light would not be visible at that distance.
			[b]Note:[/b] Using negative or values higher than [code]10.0[/code] may lead to unexpected results.
		</member>
		<member name="area_height" type="float" setter="set_param" getter="get_param" default="1.0">
			The height of the area in meters.
		</member>
		<member name="area_range" type="float" setter="set_param" getter="get_param" default="5.0">
			The range of the area in meters. This determines the maximum distance at which the area can still emit light relative to its surface, but also to its borders (which are positioned according to [member area_width] and [member area_width].
		</member>
		<member name="area_width" type="float" setter="set_param" getter="get_param" default="1.0">
			The width of the area in meters.
		</member>
		<member name="light_size" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.5" />
		<member name="light_specular" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.5" />
		<member name="shadow_bias" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="0.03" />
		<member name="shadow_normal_bias" type="float" setter="set_param" getter="get_param" overrides="Light3D" default="1.0" />
	</members>
```

CookieBadger commented:
@Calinou do you have any further details about the self-shadow bug? I was unable to reproduce it on my machine with your project.

CookieBadger commented:
@LiveTrower I would see no reason, why area lights shouldn't work on mobile, or in compatibility. In fact, they should work now (just need to fix up the shadows in compat).

LiveTrower commented:
@CookieBadger I was talking about performance because when it comes to mobile rendering and compatibility, it's a very sensitive subject. That's why many advanced features found in Forward+ rendering are left out of these renders.

You'll probably need to make a highly optimized version for these renders and prove with extensive testing that it's feasible on mobiles that support Vulkan or OpenGL, as well as on very old PC hardware that only runs Godot in compatibility mode because it lacks Vulkan or DirectX 12.

I'm not an expert, and I won't have the final say, so it would be better if @clayjohn gave you his opinion since he knows more about this.

CookieBadger commented:
@Calinou I agree, lighting computations are just mildly more complex than for point lights. Shadows in compatibility don't work since there is no dual paraboloid shadow pass either, so I disabled them now. For the mobile renderer, shadows do work, but as you say, variable penumbra doesn't. Would you rather note that in documentation, or disable shadows entirely there?

Calinou reviewed (commented):
Tested again, it works as expected.

Note that in Compatibility, the specular light lobe is missing entirely on both rough and smooth materials. This is not an issue in Forward+ and Mobile.

> [!NOTE]
>
> When testing the MRP I linked in https://github.com/godotengine/godot/pull/108219#pullrequestreview-2983649861 on Mobile and Compatibility, keep in mind the per-mesh light limits. The MRP has more than 8 area lights, which means some of them won't appear on the large ground surface until you hide some of them.

Mickeon commented:
I don't know if it's standard terminology, but I would avoid naming it **AreaLight3D** as "Area" in Godot usually refers to the entirely separate concept in the physics engine.

jcostello commented:
AreaLight is a common term to refer to this kind of lights

LiveTrower commented:
I was wondering if it's worth adding other shapes of LTC besides the rectangle one, and I think the only one that would be worthwhile is the tube or line shape since many real-world light sources have that form. The other shapes, like a disk, are very rare to see, and the sphere shape can be more or less created with an omnilight by increasing its 'size' property, but it looks more like an egg than a sphere. Is this normal?
<img width="541" height="574" alt="OmniLight" src="https://github.com/user-attachments/assets/e262072f-8a9b-4607-9f9c-5ff82a85121e" />
Also mention that with the roughness at 0 the reflection glitches.
<img width="515" height="570" alt="OmniLightBug" src="https://github.com/user-attachments/assets/621729a0-de13-40cf-9758-4b5fcf102043" />
But I also saw that you can create a line and sphere shapes without it being an area light. In fact, Unreal Engine provides [this option](https://dev.epicgames.com/documentation/en-us/unreal-engine/point-lights-in-unreal-engine) in its point light as does Wicked Engine. This could be easier to implement and maintain, as a big obstacle with LTC area lights is that their code isn't reusable. Adding another shape can mean having to create a different implementation for each shape and taking into account effects such as GI, subsurface scattering, anisotropy, clearcoat, toon shading, etc. This can complicate maintaining so many versions of area lights, and I'm not sure if the LUTs can be reused, since from what I've seen, you also have to generate a different LUT for each shading effect.

LiveTrower commented on servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:
```diff
@@ -909,6 +909,407 @@ void light_process_spot(uint idx, vec3 vertex, hvec3 eye_vec, hvec3 normal, vec3
+		diffuse_light += ltc_diffuse * color / half(2.0 * M_PI) * light_attenuation;
+	}
+	hvec3 spec = ltc_specular * color;
+	hvec3 spec_color = mix(hvec3(0.04), albedo, hvec3(metallic));
```
Is there a reason to use this formula instead of the existing **f0**?
It would also be good to include **f90** to maintain the same behavior as the other types of light.

Calinou commented:
Left: Emissive plane pointing towards the ground with black albedo, emission energy set to `15.0`.
Right (from top to bottom): AreaLight3D with bake modes Disabled, Static, Dynamic. Energy is `3.0` on all of them.

Denoising is disabled, bake quality is set to Ultra and supersampling is enabled with a `2.0` factor.

Directional OFF | Directional ON
-|-
<img width="648" height="648" alt="Screenshot_20250826_175416" src="https://github.com/user-attachments/assets/c59489b0-08d4-44e0-bcf8-c8f6ab8d901a" /> | <img width="648" height="648" alt="Screenshot_20250826_175422" src="https://github.com/user-attachments/assets/4e71218c-1660-4f86-ad9e-f10a49ecd240" />

It seems quite accurate, although you need to use a much lower energy compared to an emissive plane to get the same result. Directional information also appears significantly less strong (not sure if that's intended).

**Testing project:** [test_area_lightmapgi.zip](https://github.com/user-attachments/files/21991319/test_area_lightmapgi.zip)

CookieBadger commented:
@QbieShay I made it work with volumetric fog now. 

The illumination adjusts to the shape of the light, which is calculated via the rect's solid angle, instead of LTCs, since the froxels don't have a *normal*.

<img width="1609" height="954" alt="image" src="https://github.com/user-attachments/assets/a34c2216-37fb-4871-b00f-449a5083617e" />

![volumetric_fog2_comp](https://github.com/user-attachments/assets/bac6ce4a-5540-418d-822b-c117677c3049)

![volumetric_fog_comp](https://github.com/user-attachments/assets/bc1a8bc5-41c6-4966-be21-6dc43cea5171)

[AreaLightsVolumetricFogTest.zip](https://github.com/user-attachments/files/22027369/AreaLightsVolumetricFogTest.zip)

Calinou commented:
Volumetric fog looks great 🙂 I've tested it on Metal and it works as expected.

However, it seems to look quite weak in comparison to an OmniLight3D with the same range, color, attenuation and volumetric fog energy factor:

<img width="999" height="543" alt="Screenshot 2025-08-28 at 17 03 46" src="https://github.com/user-attachments/assets/475d22e2-3e72-4821-96e0-809fbf16262e" />

The AreaLight3D has a size of `(2, 2)` and I've maxed out Volume Size/Volume Depth project settings to give the most accurate representation possible.

CookieBadger commented:
Thanks for testing, I confused a conditional operator; fixed. 
One thing I forgot to mention is that right now, the light needs to have a minimum range of 1.0m to affect fog at all. Otherwise we get intense flickering at the surface of the light.

As to why the light is weak compared to the point light: I divided the solid angle by 2*PI to give a ratio of how much of the hemisphere is covered by the light. Physically, this should make sense. But, if we remove this division, we get a light that is stronger and more similar in its intensity to the point light at a size of `(1,1)`. That change would also make the area light much stronger compared to point lights when it has a large size.

Btw the energy of an area light is interpreted as illuminance (luminous power per unit area), i.e. if you increase the size of the light, the illumination of the scene is stronger. A different approach would be to interpret the energy parameter as power, i.e. divide it by the area before using it in lighting calculations.

jcostello commented:
> @Calinou Thanks for testing, I confused a conditional operator; fixed. One thing I forgot to mention is that right now, the light needs to have a minimum range of 1.0m to affect fog at all. Otherwise we get intense flickering at the surface of the light.
> 
> As to why the light is weak compared to the point light: I divided the solid angle by 2*PI to give a ratio of how much of the hemisphere is covered by the light. Physically, this should make sense. But, if we remove this division, we get a light that is stronger and more similar in its intensity to the point light at a size of `(1,1)`. That change would also make the area light much stronger compared to point lights when it has a large size.
> 
> Btw the energy of an area light is interpreted as illuminance (luminous power per unit area), i.e. if you increase the size of the light, the illumination of the scene is stronger. A different approach would be to interpret the energy parameter as power, i.e. divide it by the area before using it in lighting calculations.

I think now blender has a normalization for lights so size don't affect the energy output. Maybe that is more intuitive for the general public

CookieBadger commented:
From my quick research, it seems area lights are unsupported in the common 3D formats. I found no existing or planned extension for area lights in the glTF format (as the name suggests, `KHR_lights_punctual` cares only about point/spot lights). fbx apparently knows about area lights, but ignores the shape and intensity. Blender, of course, stores area lights with an enum for its shape, and a size field (https://docs.blender.org/api/current/bpy.types.AreaLight.html). 

I did also look at physical light units. Blender stores all lights in terms of radiant power (`W` = Watts), which is non-standard. 
When exporting Blender point lights to glTF in "Standard" units, watts are converted to candela (`= lm/sr`), with the conversion `S / (4π) * 683`, see https://github.com/godotengine/godot/issues/73624#issuecomment-1745578491.
The energy of area lights can be represented either in lumens (luminous power), or nits (luminance, cd/m<sup>2</sup>), depending on whether a change in size should affect energy output. So, the Blender lights would be converted to lumens with the conversion `S * 683`, but I remember when matching light intensities between Blender and Gotot empirically, I had a very different conversion rate. 
As for other engines, Unity allows selecting between lumens and nits, with lumens being the default. If nits are selected, changes in size do affect energy output. Unreal Engine allows selecting between candelas and lumens, with candelas being the default (1cd=3,14 lm, which for my understanding does not make much sense, since the solid angle of a hemisphere is 2π sr, so it should be 1cd = 6,14lm for an area light?). In any case, changing size does not affect energy output. Both engines also support EV. 

-> It seems, that the more easily understood unit is more likely lumens and that size should not affect energy output. I should be able to adjust for that when I have time.

Calinou commented:
I've tried the latest revision of the PR with light normalization based on the area size, and I'm not sure it's better in all scenarios. The light visibly gets a lot weaker as you increase its size. You'll need very high energy values like `100` to make the light strong enough at large sizes like `10x10`. Maybe it'll need to be a toggleable property (`false` by default), similar to glow where you can choose whether to normalize levels.

@passivestar @lander-vr I'm curious about your opinion on this topic.

passivestar commented:
> I'm curious about your opinion on these topics.

I agree, normalization may or may not be useful based on what you're trying to achieve so a toggle makes sense. Blender for example has it:

<img width="470" height="102" alt="image" src="https://github.com/user-attachments/assets/d3462153-13bb-4756-aebf-ae3b63d777e7" /><br>

> Maybe it'll need to be a toggleable Normalize Energy property in the Area section (false by default)

For reference, blender has it on by default, and it also normalizes energy of point lights by default, unlike godot (if you make a point light's size bigger, it becomes dimmer)

> While looking at this, I was also thinking whether the Light > Size property should be relative to the area size. At small area sizes, shadows are very hard to spot as the default light size (0.5) is very high relative to the area size.

I wasn't following this thread until the ping so I don't know if this was discussed, but from the user perspective it seems confusing that there are now 2 size properties. I would expect the new area light size prop to take care of shadows too. If this is just a tech debt thing and we're stuck with two different size properties for the light and the shadow until Godot 5, then I wouldn't make one dependent on another and keep them separate to avoid making things even more confusing 😁

Also I just tested the PR and found that the old light size property affects the reflection, it probably shouldn't be happening in an area light?

https://github.com/user-attachments/assets/09d0b38e-7d95-438b-87c7-62e5ff0fe36a

CookieBadger commented:
So.... I added the checkbox for the normalization now. I also realized while investigating the fog issue, that attenuation was calculated incorreclty. Just so you know, area lights (calculated with LTC or with solid angle) already attenuate with the inverse square by default. I used to multiply by the omni attenuation to the closest point, so we used to have a cubic attenuation, which is why the values were so different to omni lights. I fixed this by calling the omni attenuation function with `attenuation - 2.0`. This means, that with a default value of 1.0 for attenuation, we actually first calculate area lights with inverse square attenuation, and then multiply by distance, such that we sort of get linear attenuation. Now as you can see in the images below, the intensities are now closely matched. 

The omni/spot lights (left) and area lights (right) below always use the same energy values.

![AreaLightAttenuation (1)](https://github.com/user-attachments/assets/8e7c851e-5750-458e-8fe9-5ba6115c13c3)

![LightEffects](https://github.com/user-attachments/assets/5e8e35c1-d9a5-424a-9b6c-8f65cbd56da4)

For fog, where I use the solid angle calculation, I actually also had to include a normalization factor of `7.801015826317776`. This magic number is calculated as `1/s`, where `s` is the solid angle of a quad with size `(1,1)` at a distance of `1` divided by `2*pi`.

Calinou commented:
Shadows work now. However, I noticed that if you change the area size but not the area range, the shadow map won't update, which leads to shadow acne/peter-panning issues:

https://github.com/user-attachments/assets/7728a9a7-3b88-41f2-8152-6a5756caf524

If you uncomment the line that always changes `area_range` by a small offset every frame in `area_light_3d.gd`, the issue goes away.

**Updated testing project:** [test_area_light_small_area_2.zip](https://github.com/user-attachments/files/22162174/test_area_light_small_area_2.zip)

CookieBadger commented:
I'll investigate the shadow bug another time, I think it has been there for a while and wasn't caused by a recent change. 
I did just finished up VoxelGI integration:

<img width="1540" height="815" alt="image" src="https://github.com/user-attachments/assets/ebb3a10a-2a0d-4e74-8e4a-fc29714b9a86" />

I was however not able to closely match the intensity of a spotlight and the area light. Maybe you can help me test it to find which parameters make the differences better or worse. The stuff I tried didn't work.

<img width="1514" height="360" alt="image" src="https://github.com/user-attachments/assets/95a29b09-6fe7-4646-bb8a-0df05fde4045" />

CookieBadger commented:
shadow bug fixed.

CookieBadger commented:
rebased.

Calinou commented:
While testing a screen-space contact shadows shader, I noticed `light()` doesn't run in custom shaders when using an AreaLight3D node, unlike other light types. How difficult would it be to implement this? (I'd say this can be left for a future PR if it's nontrivial.)

You can test this with this function, which should tint the whole scene red whenever there's any light in view:

```glsl
void light() {
	DIFFUSE_LIGHT += vec3(1.0, 0.0, 0.0);
}
```

fire commented:
@CookieBadger Are you able to look into @Calinou 's question about `light()` in custom shaders.

fire commented:
Did you try  ERROR: /root: The caller thread can't call the function `propagate_notification()` on this node. Use `call_deferred()` or `call_deferred_thread_group()` instead.
   at: propagate_notification (./scene/main/node.cpp:2597)?

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2893,6 +2927,258 @@ void TextureStorage::_texture_format_from_rd(RD::DataFormat p_rd_format, Texture
+					AreaLightAtlas::Texture *t = area_light_atlas.textures.getptr(E.key);
+					Texture *src_tex = get_texture(E.key);
+					Rect2 uv_rect = t->uv_rect;
+					if (MIN(uv_rect.size.x * area_light_atlas.size.x, uv_rect.size.y * area_light_atlas.size.y) > pow(2, i)) {
```
```suggestion
					if (MIN(uv_rect.size.x * area_light_atlas.size.x, uv_rect.size.y * area_light_atlas.size.y) > Math::pow(2, i)) {
```

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2893,6 +2927,258 @@ void TextureStorage::_texture_format_from_rd(RD::DataFormat p_rd_format, Texture
+					Texture *src_tex = get_texture(E.key);
+					Rect2 uv_rect = t->uv_rect;
+					if (MIN(uv_rect.size.x * area_light_atlas.size.x, uv_rect.size.y * area_light_atlas.size.y) > pow(2, i)) {
+						Vector2i mip_size = area_light_atlas.size / pow(2, i);
```
```suggestion
						Vector2i mip_size = area_light_atlas.size / Math::pow(2, i);
```

AThousandShips commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1636,6 +1648,66 @@ TypedArray<Image> RendererSceneRenderRD::bake_render_uv2(RID p_base, const Typed
+	PackedByteArray data;
+	ERR_FAIL_COND_V_MSG(p_mipmaps <= 0, data, "Mipmaps must be greater than 0");
+	ERR_FAIL_COND_V_MSG(p_size.width < pow(2, p_mipmaps), data, "Image width must be greater than mipmaps to power of 2");
+	ERR_FAIL_COND_V_MSG(p_size.height < pow(2, p_mipmaps), data, "Image height must be greater than mipmaps to power of 2");
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_size.width < Math::pow(2, p_mipmaps), data, "Image width must be greater than mipmaps to power of 2.");
	ERR_FAIL_COND_V_MSG(p_size.height < Math::pow(2, p_mipmaps), data, "Image height must be greater than mipmaps to power of 2.");
```

AThousandShips commented on scene/3d/lightmapper.h:
```diff
@@ -170,6 +171,8 @@ class Lightmapper : public RefCounted {
 	virtual void add_directional_light(const String &p_name, bool p_static, const Vector3 &p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_angular_distance, float p_shadow_blur) = 0;
 	virtual void add_omni_light(const String &p_name, bool p_static, const Vector3 &p_position, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_size, float p_shadow_blur) = 0;
 	virtual void add_spot_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_spot_angle, float p_spot_attenuation, float p_size, float p_shadow_blur) = 0;
+	virtual void add_area_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, const Vector3 &p_area_width, const Vector3 &p_area_height, float p_size, float p_shadow_blur, const Rect2 &p_texture_rect) = 0;
```
```suggestion
	virtual void add_area_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 &p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, const Vector3 &p_area_width, const Vector3 &p_area_height, float p_size, float p_shadow_blur, const Rect2 &p_texture_rect) = 0;
```

AThousandShips commented on scene/3d/lightmapper.h:
```diff
@@ -170,6 +171,8 @@ class Lightmapper : public RefCounted {
 	virtual void add_omni_light(const String &p_name, bool p_static, const Vector3 &p_position, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_size, float p_shadow_blur) = 0;
 	virtual void add_spot_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_spot_angle, float p_spot_attenuation, float p_size, float p_shadow_blur) = 0;
+	virtual void add_area_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, const Vector3 &p_area_width, const Vector3 &p_area_height, float p_size, float p_shadow_blur, const Rect2 &p_texture_rect) = 0;
+	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &atlas_data) = 0;
```
```suggestion
	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &p_atlas_data) = 0;
```

AThousandShips replied:
```suggestion
	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &p_atlas_data) = 0;
```

AThousandShips commented on scene/3d/lightmap_gi.cpp:
```diff
@@ -893,6 +896,138 @@ LightmapGI::BakeError LightmapGI::_save_and_reimport_atlas_textures(const Ref<Li
+
+	for (int i = 0; i < item_count; i++) {
+		Rect2 uv_rect;
+		uv_rect.position = items[i].pos * border + Vector2i(border / 2, border / 2);
```
```suggestion
		uv_rect.position = items[i].pos * border + Vector2i(border, border) / 2;
```

AThousandShips commented on scene/3d/lightmap_gi.cpp:
```diff
@@ -893,6 +896,138 @@ LightmapGI::BakeError LightmapGI::_save_and_reimport_atlas_textures(const Ref<Li
 
+void LightmapGI::_build_area_light_texture_atlas(const Vector<LightmapGI::LightsFound> &lights_found, HashMap<Ref<Texture2D>, Rect2> &r_texture_rects, Size2i &r_atlas_size, int &r_mipmaps) const {
+	r_mipmaps = 8;
+	r_atlas_size = Size2i(pow(2, r_mipmaps), pow(2, r_mipmaps));
```
```suggestion
	r_atlas_size = Size2i(Math::pow(2, r_mipmaps), Math::pow(2, r_mipmaps));
```

AThousandShips replied:
```suggestion
	r_atlas_size = Size2i(Math::pow(2, r_mipmaps), Math::pow(2, r_mipmaps));
```

AThousandShips commented on modules/lightmapper_rd/lightmapper_rd.h:
```diff
@@ -302,6 +311,8 @@ class LightmapperRD : public Lightmapper {
 	virtual void add_omni_light(const String &p_name, bool p_static, const Vector3 &p_position, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_size, float p_shadow_blur) override;
 	virtual void add_spot_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, float p_spot_angle, float p_spot_attenuation, float p_size, float p_shadow_blur) override;
+	virtual void add_area_light(const String &p_name, bool p_static, const Vector3 &p_position, const Vector3 p_direction, const Color &p_color, float p_energy, float p_indirect_energy, float p_range, float p_attenuation, const Vector3 &p_area_width, const Vector3 &p_area_height, float p_size, float p_shadow_blur, const Rect2 &p_texture_rect) override;
+	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &atlas_data) override;
```
```suggestion
	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &p_atlas_data) override;
```

AThousandShips replied:
```suggestion
	virtual void add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &p_atlas_data) override;
```

AThousandShips commented on modules/lightmapper_rd/lightmapper_rd.cpp:
```diff
@@ -139,6 +139,49 @@ void LightmapperRD::add_spot_light(const String &p_name, bool p_static, const Ve
+	light_metadata.push_back(md);
+}
+
+void LightmapperRD::add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &atlas_data) {
```
```suggestion
void LightmapperRD::add_area_light_atlas(const Vector2i &p_size, int p_mipmap_count, const PackedByteArray &p_atlas_data) {
```

jcostello commented:
And in other projects

```
SPIR-V Parsing error. Result 3803 is not valid.
SPIR-V Parsing error. Result 3790 is not valid.
SPIR-V Parsing error. Result 3773 is not valid.
SPIR-V Parsing error. Result 3786 is not valid.

WARNING: "usage" is not supported in add_property_info().
     at: _add_property_info_bind (core/config/project_settings.cpp:1347)
WARNING: "usage" is not supported in add_property_info().
     at: _add_property_info_bind (core/config/project_settings.cpp:1347)
ERROR: /root: The caller thread can't call the function `propagate_notification()` on this node. Use `call_deferred()` or `call_deferred_thread_group()` instead.
   at: propagate_notification (scene/main/node.cpp:2597)
```

Calinou commented:
I get a crash in Forward+ when enabling Subsurf Scatter Transmittance (enabling Subsurf Scatter alone is fine):

```
[...]
4986 |  fragment_shader(scene_data_block.data);
4987 | }
4988 | 
4989 | 
   at: compile_stages (./servers/rendering/renderer_rd/shader_rd.cpp:1004)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Condition "variant_stages.is_empty()" is true.
   at: _compile_variant (./servers/rendering/renderer_rd/shader_rd.cpp:306)
ERROR: Parameter "shader" is null.
   at: uniform_set_create (./servers/rendering/rendering_device.cpp:3667)
ERROR: Parameter "us" is null.
   at: uniform_set_set_invalidation_callback (./servers/rendering/rendering_device.cpp:4087)
ERROR: FATAL: Index p_index = 3 is out of bounds (size() = 0).
   at: get (./core/templates/cowdata.h:193)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.6.dev.custom_build (bf0fd89082e2757c0b6cc5aef6a9030b64260bf2)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x1a2c0) [0x7f41a2b322c0] (??:0)
[2] gde() [0x7ab05a7] (/home/hugo/Documents/Git/godotengine/godot/core/templates/cowdata.h:193 (discriminator 13))
[3] gde() [0x7ab5cf7] (/home/hugo/Documents/Git/godotengine/godot/core/templates/cowdata.h:170)
[4] gde() [0x7b5eb26] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:4288 (discriminator 3))
[5] gde() [0x7b5f55b] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:4342 (discriminator 1))
[6] gde() [0x7b7d94c] (/home/hugo/Documents/Git/godotengine/godot/core/templates/self_list.h:127)
[7] gde() [0x7ba146b] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:693 (discriminator 1))
[8] gde() [0x79e1ba8] (/home/hugo/Documents/Git/godotengine/godot/core/templates/paged_array.h:148)
[9] gde() [0x77cd635] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_scene_cull.cpp:3578 (discriminator 1))
[10] gde() [0x77d600f] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_scene_cull.cpp:2772)
[11] gde() [0x7779205] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_viewport.cpp:320)
[12] gde() [0x7794e84] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_viewport.cpp:387)
[13] gde() [0x7916e5c] (/home/hugo/Documents/Git/godotengine/godot/servers/rendering/renderer_viewport.cpp:896)
[14] gde() [0x379b167] (/home/hugo/Documents/Git/godotengine/godot/core/config/engine.cpp:413)
[15] gde() [0x3764e2b] (/home/hugo/Documents/Git/godotengine/godot/platform/linuxbsd/os_linuxbsd.cpp:1002 (discriminator 1))
[16] /lib64/libc.so.6(+0x35b5) [0x7f41a2b1b5b5] (??:0)
[17] /lib64/libc.so.6(__libc_start_main+0x88) [0x7f41a2b1b668] (??:0)
[18] gde() [0x35fd065] (??:?)
-- END OF C++ BACKTRACE --
================================================================
```

This happens even if the scene doesn't have any AreaLight3D nodes, but it doesn't occur on `master`.

Opening the editor on a new project in Mobile also results in shader compilation error spam:

```
ERROR: Uniforms supplied for set (0):
Set: 0 Binding: 1 Type: StorageBuffer Writable: N Length: 0
Set: 0 Binding: 2 Type: UniformBuffer Writable: N Length: 352
[...]
Set: 0 Binding: 14 Type: Sampler Writable: N Length: 1
Set: 0 Binding: 15 Type: Sampler Writable: N Length: 1
are not the same format as required by the pipeline shader. Pipeline shader requires the following bindings:
Set: 0 Binding: 2 Type: Sampler Writable: N Length: 1
Set: 0 Binding: 3 Type: StorageBuffer Writable: N Length: 0
[...]
Set: 2 Binding: 0 Type: StorageBuffer Writable: N Length: 0
Set: 3 Binding: 0 Type: UniformBuffer Writable: N Length: 16
   at: draw_list_draw (./servers/rendering/rendering_device.cpp:5004)
```

___

Some videos of area light textures in action:

Real-time texture updates with shadows:

[area_lights.webm](https://github.com/user-attachments/assets/4cbdda94-e363-420d-a896-1f4723ea0cf9)

Video playback using low-resolution textures (created from the OGV file with FFmpeg), changed every frame:

[area_light_video.webm](https://github.com/user-attachments/assets/795c020f-2483-4e89-9822-819f5fb241b1)

Better stability could be achieved by blurring the frames in FFmpeg.

coobenguy commented:
Really great work, the only issue Ive found is textures 128x128 and below dont display correctly 
<img width="2451" height="1320" alt="image" src="https://github.com/user-attachments/assets/693c4b6b-cefd-482a-822c-d666cabe34ea" />
<img width="2099" height="1292" alt="image" src="https://github.com/user-attachments/assets/d77a153d-e087-45dd-93fc-9299c1b736cc" />

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2893,6 +2927,285 @@ void TextureStorage::_texture_format_from_rd(RD::DataFormat p_rd_format, Texture
+			Texture *src_tex = get_texture(E.key);
+
+			si.size.width = ceil(float(src_tex->width) / border);
+			si.size.height = ceil(float(src_tex->height) / border);
```
```suggestion
			si.size.width = Math::ceil(float(src_tex->width) / border);
			si.size.height = Math::ceil(float(src_tex->height) / border);
```

AThousandShips commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1636,6 +1648,66 @@ TypedArray<Image> RendererSceneRenderRD::bake_render_uv2(RID p_base, const Typed
+			Rect2 uv_rect = p_area_light_atlas_texture_rects[t_idx];
+			uv_rect.position = CLAMP(uv_rect.position, Vector2(0.0, 0.0), Vector2(1.0, 1.0));
+			uv_rect.size = CLAMP(uv_rect.size, Vector2(0.0, 0.0), Vector2(1.0 - uv_rect.position.x, 1.0 - uv_rect.position.y));
+			Vector2i mip_size = p_size / pow(2, i);
```
```suggestion
			Vector2i mip_size = p_size / Math::pow(2, i);
```

AThousandShips commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1636,6 +1648,66 @@ TypedArray<Image> RendererSceneRenderRD::bake_render_uv2(RID p_base, const Typed
+			RID texture = p_area_light_textures[t_idx];
+			Rect2 uv_rect = p_area_light_atlas_texture_rects[t_idx];
+			uv_rect.position = CLAMP(uv_rect.position, Vector2(0.0, 0.0), Vector2(1.0, 1.0));
+			uv_rect.size = CLAMP(uv_rect.size, Vector2(0.0, 0.0), Vector2(1.0 - uv_rect.position.x, 1.0 - uv_rect.position.y));
```
```suggestion
			uv_rect.position = uv_rect.position.clampf(0.0, 1.0);
			uv_rect.size = uv_rect.size.clamp(Vector2(0.0, 0.0), Vector2(1.0 - uv_rect.position.x, 1.0 - uv_rect.position.y));
```
Unless you mean clamping by the full vector, not element wise

CookieBadger commented:
@jcostello Do you have any particular problems with physical light? As far as I can tell it works, but my knowledge about that feature is limited.

jcostello commented:
Still having problems baking lightmaps. MRP

[reflection-lightmap.zip](https://github.com/user-attachments/files/24154841/reflection-lightmap.zip)

AThousandShips commented on servers/rendering/rendering_server.h:
```diff
@@ -532,7 +532,8 @@ class RenderingServer : public Object {
 		LIGHT_OMNI,
-		LIGHT_SPOT
+		LIGHT_SPOT,
+		LIGHT_AREA
```
```suggestion
		LIGHT_AREA,
```

AThousandShips commented on servers/rendering/renderer_rd/cluster_builder_rd.h:
```diff
@@ -334,6 +339,39 @@ class ClusterBuilderRD {
+				float s = xform.basis.rows[i].length();
+				//scale[i] *= s; // lights ignore scale
+				xform.basis.rows[i] /= s;
+			};
```
```suggestion
			}
```

AThousandShips commented on doc/classes/AreaLight3D.xml:
```diff
@@ -0,0 +1,31 @@
+		An area light is a type of [Light3D] node that emits light over a 2-dimensional area, in the shape of a rectangle. The light is attenuated through the distance. This attenuation can be configured by changing the energy, [member area_attenuation], and [member area_range].
+		Light is emitted in the -Z direction of the node's global basis. For an unrotated light, this means that the light is emitted forwards, illuminating the front side of a 3D model (see [constant Vector3.FORWARD] and [constant Vector3.MODEL_FRONT]).
+		Area lights can cast soft shadows using PCSS, which you can control by tweaking the size parameter. The shadow map is drawn from the center of the light.
+		[b]Note:[/b] Area lights have limited support in the Mobile or Compatibility renderer. In the Mobile renderer, the size of the penumbra doesn't vary is it should with PCSS. In Compatibility, area lights cannot cast shadows. For static lighting, consider instead using [LightmapGI] and a [QuadMesh] with [member PrimitiveMesh.add_uv2] set to [code]true[/code] and assign an emissive material to the ...
```
```suggestion
		[b]Note:[/b] Area lights have limited support in the Mobile and Compatibility renderers. In the Mobile renderer, the size of the penumbra doesn't vary as it should with PCSS. In Compatibility, area lights cannot cast shadows. For static lighting, consider instead using [LightmapGI] and a [QuadMesh] with [member PrimitiveMesh.add_uv2] set to [code]true[/code], assign an emissive material to the [QuadMesh], then hide the [QuadMesh] after baking lightmaps.
```

AThousandShips commented on doc/classes/AreaLight3D.xml:
```diff
@@ -0,0 +1,34 @@
+			[b]Note:[/b] Using negative or values higher than [code]10.0[/code] may lead to unexpected results.
+		</member>
+		<member name="area_normalize_energy" type="bool" setter="set_area_normalize_energy" getter="get_area_normalize_energy" default="true">
+			Defines whether the energy is normalized (divided) by the surface area of the light. If set to [code]true[/code], changing the size does not affect the amount of total energy output, and does not dramatically alter the brightness of the scene.
```
```suggestion
			Defines whether the energy is normalized (divided) by the surface area of the light. If set to [code]true[/code], changing the size does not affect the total energy output, and does not dramatically alter the brightness of the scene.
```

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -2893,6 +2927,292 @@ void TextureStorage::_texture_format_from_rd(RD::DataFormat p_rd_format, Texture
+			AreaLightAtlas::SortItem &si = itemsv.write[idx];
+
+			Texture *src_tex = get_texture(E.key);
+			Vector2i b_size = Vector2i(ceil(float(src_tex->width) / border), ceil(float(src_tex->height) / border));
```
```suggestion
			Vector2i b_size = Vector2i(Math::ceil(float(src_tex->width) / border), Math::ceil(float(src_tex->height) / border));
```

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/light_storage.cpp:
```diff
@@ -945,6 +1128,17 @@ void LightStorage::update_light_buffers(RenderDataRD *p_render_data, const Paged
+			light_data.projector_rect[2] = rect.size.width;
+			light_data.projector_rect[3] = rect.size.height;
+			Size2i texture_size = (rect.size * texture_storage->area_light_atlas_get_size()).ceil();
+			light_data.cos_spot_angle = MIN(floor(log2(MAX(MIN(texture_size.x, texture_size.y), 1)) - 1.0), texture_storage->area_light_atlas_get_mipmaps()); // max mipmaps
```
```suggestion
			light_data.cos_spot_angle = MIN(Math::floor(Math::log2(MAX(MIN(texture_size.x, texture_size.y), 1)) - 1.0), texture_storage->area_light_atlas_get_mipmaps()); // max mipmaps
```

AThousandShips commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -1636,6 +1648,86 @@ TypedArray<Image> RendererSceneRenderRD::bake_render_uv2(RID p_base, const Typed
+		RID texture = p_area_light_textures[t_idx];
+		Rect2 uv_rect = p_area_light_atlas_texture_rects[t_idx];
+		uv_rect.position = CLAMP(uv_rect.position, Vector2(0.0, 0.0), Vector2(1.0, 1.0));
+		uv_rect.size = CLAMP(uv_rect.size, Vector2(0.0, 0.0), Vector2(1.0 - uv_rect.position.x, 1.0 - uv_rect.position.y));
```
```suggestion
		uv_rect.position = uv_rect.position.clampf(0.0, 1.0);
		uv_rect.size = uv_rect.size.clamp(Vector2(0.0, 0.0), Vector2(1.0 - uv_rect.position.x, 1.0 - uv_rect.position.y));
```
Unless you mean clamping by the full vector, not element wise

AThousandShips commented on doc/classes/RenderingServer.xml:
```diff
@@ -18,6 +18,11 @@
 	<methods>
+		<method name="area_light_create">
+			<return type="RID" />
+			<description>
```
Needs a description

jcostello commented:
Also, static intensity seems to be lower than realtime (Not using a light texture)

Static bake
<img width="1136" height="793" alt="image" src="https://github.com/user-attachments/assets/f2f4d25e-97a1-4ace-9ff9-2614eea53142" />

Real Time
<img width="1136" height="793" alt="image" src="https://github.com/user-attachments/assets/13d7e33b-1363-40a7-adff-831aede2d8ec" />

Calinou commented on doc/classes/AreaLight3D.xml:
```diff
@@ -0,0 +1,37 @@
+		An area light is a type of [Light3D] node that emits light over a two-dimensional area, in the shape of a rectangle. The light is attenuated throughout the distance. This attenuation can be configured by changing the energy, [member area_attenuation], and [member area_range].
+		Light is emitted in the -Z direction of the node's global basis. For an unrotated light, this means that the light is emitted forwards, illuminating the front side of a 3D model (see [constant Vector3.FORWARD] and [constant Vector3.MODEL_FRONT]).
+		Area lights can cast soft shadows using PCSS, which you can control by tweaking the size parameter. The shadow map is drawn from the center of the light.
+		[b]Note:[/b] Area lights have limited support in the Mobile and Compatibility renderers. In the Mobile renderer, the size of the penumbra doesn't vary as it should with PCSS. In Compatibility, area lights cannot cast shadows. For static lighting, consider instead using [LightmapGI] and a [QuadMesh] with [member PrimitiveMesh.add_uv2] set to [code]true[/code], assign an emissive material to the  ...
```
This workaround suggestion can be removed, now that lightmaps support AreaLight3D in all renderers:

```suggestion
		[b]Note:[/b] Area lights have limited support in the Mobile and Compatibility renderers. In the Mobile renderer, the size of the penumbra doesn't vary as it should with PCSS. In Compatibility, area lights cannot cast shadows.
```

(It's technically more efficient to use the workaround, as it will disable direct light receiving on dynamic objects, but this would be better addressed by https://github.com/godotengine/godot-proposals/issues/8040.)

jcostello commented:
I just tested it and attenunation works close enough to real time. The only issue I just found is that shadows works diferently when baked static vs real time depending of the light size

Realtime
<img width="1073" height="657" alt="Screenshot From 2025-12-16 10-11-45" src="https://github.com/user-attachments/assets/7aba4f6f-870a-41ce-bd38-b912937cc642" />

Static
<img width="1073" height="657" alt="Screenshot From 2025-12-16 10-11-50" src="https://github.com/user-attachments/assets/325db981-b4ba-4a8f-9fd5-665194c1dfdd" />

<img width="253" height="630" alt="Screenshot From 2025-12-16 10-13-30" src="https://github.com/user-attachments/assets/5875265d-c3a8-4b90-a3b1-5152a40d9f9e" />

CookieBadger commented:
@jcostello fixed. We could actually implement a monte-carlo ray-tracing loop there for accurate soft shadows, but that would be a topic for future work.
<img width="881" height="462" alt="image" src="https://github.com/user-attachments/assets/e81f2664-ac01-4596-94f2-c87aec1bc021" />

jcostello commented:
@CookieBadger works as intended. I agree that implementing monte-carlo ray-tracing is for a future PR. Go ahead an create a proposal for it

CookieBadger commented:
yeah without an MRP it is not possible to tell just from the screenshots what is actually wrong, e.g. what kind of lights are you baking, what are your baking settings, and I am not clear what the issue in the second image is.

CookieBadger commented:
<img width="1069" height="1188" alt="image" src="https://github.com/user-attachments/assets/78dd1487-a772-4dc8-afa7-64c6f2b9e1b9" />
I see now what leak you are describing. This is happening due to the inaccuracies of the dual paraboloid projections, the same happens when using point lights with DP shadows. This is actually described in the description of the PR. I'm not sure if improving the accuracies as described will prevent a leak as bad as that, but know that the PCSS area light shadows are an crude approximation that must be used with care. Without ray-tracing, accurate area light shadows are not possible. In the scenario we have here, proper handling would be setting an appropriate range for the light, such that it does not illuminate through the ground. 

My lightmaps build correctly, but again, I don't know what settings you are using.

wagnerfs commented:
@Calinou You're right, I deleted my other post because I didn't notice the depth subdivision was 0.
Subdividing does the trick.

clayjohn reviewed (changes requested):
Lots of little comments from the rendering meeting review!

LiveTrower commented on servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl:
```diff
@@ -928,9 +946,366 @@ void light_process_spot(uint idx, vec3 vertex, hvec3 eye_vec, hvec3 normal, vec3
+	float cc_specular_ltc = 0.0;
+	vec2 cc_fresnel;
+	ltc_evaluate_specular(vec3(vertex_normal), vec3(eye_vec), sqrt(mix(0.001, 0.1, float(clearcoat_roughness))), points, area_lights.data[idx].projector_rect, max_mipmap, area_light_atlas, light_projector_sampler, ltc_lut1, ltc_lut2, cc_specular_ltc, cc_fresnel, cc_specular_tex_color);
+	half Fr = half(0.04) * max(half(cc_fresnel.x), half(0.0)) + half(1.0 - 0.04) * max(half(cc_fresnel.y), half(0.0)) * clearcoat;
```
```suggestion
	half Fr = (half(0.04) * max(half(cc_fresnel.x), half(0.0)) + half(1.0 - 0.04) * max(half(cc_fresnel.y), half(0.0))) * clearcoat;
```
The `clearcoat` parameter doesn't reduce the clearcoat effect; to do that, you need to use parentheses to prioritize the Fresnel calculation and then multiply it by the parameter.

CookieBadger commented:
Implemented area lights in SDFGI. For textures to work, I had to add another uniform set in `gi.h: SDFGI`, which is updated per frame. The algorithm should be tweaked in the future, as it currently does not match point lights that well, especially when textures are used.

Calinou reviewed (commented):
Tested locally, it works as expected with all renderers. I've also tested SDFGI and volumetric fog.

Documentation looks good to me.

Some feedback:

- The **Projector** property is visible in the AreaLight3D inspector, but it has no effect (even when shadows are enabled). The Texture property is used instead.

Calinou commented:
I found one last issue: depending on the value of the light projectors project setting, area light texture filtering may look broken:

https://github.com/user-attachments/assets/fd4ae596-fa1a-4614-ba51-9d633e89e8bb

Values with mipmaps enabled are fine (**Nearest** values will make reflection textures look crisp, which is good for pixel art), but values without mipmaps don't render correctly.

Resolving this issue will probably require adding a dedicated sampler for area light textures (which is configured to always use mipmaps), or at least documenting this limitation. I'm also not sure if anisotropic filtering is really taken advantage of for light projector textures. If it isn't, it's a good opportunity to ignore it in the new sampler and always go for isotropic filtering to save on memory bandwidth.

Changing **Texture Mipmap Bias** will also affect area light rendering, but we already have this issue with reflection probes, so it's probably fine to address later:

https://github.com/user-attachments/assets/3c7488c2-38ed-48ab-bb96-9a4ac0ce0e7d

**Testing project:** [test_pr_108219_filter.zip](https://github.com/user-attachments/files/25640786/test_pr_108219_filter.zip)

Calinou reviewed (approved):
I did some final testing on all renderers with D3D12 (including WARP software rendering), seems fine. Code looks good to me.

I've spotted some small optimizations that can be applied to the area light shader, but it's probably easier to perform those in a future PR so I can also benchmark before/after.

passivestar commented:
Tested, looks great to me, didn't see any issues

<img width="2047" alt="image" src="https://github.com/user-attachments/assets/63038fae-2db0-4884-a334-5a55d2b2f6b4" />

QbieShay commented:
Still in the process of reviewing, some feedback for now:

1. the documentation of the `light_size` property needs to be updated with matching description for area lights.
2. I think i'm in a cursed scenario, but shadows look kind of strange
<img width="883" height="552" alt="image" src="https://github.com/user-attachments/assets/a93957b4-c9d0-43ec-bf60-57666b7f4d7c" />
<img width="766" height="441" alt="image" src="https://github.com/user-attachments/assets/2fb591cb-655b-48eb-8d4e-c7a0e92ba733" />

Omni light does't seem to have the same problem
<img width="600" height="439" alt="image" src="https://github.com/user-attachments/assets/2d92b78f-e8e9-43b7-9c86-d582efe3e116" />

3. Fog seems to be overall alright. There are some edge cases that puzzle me a bit, but i can't tell whether this is expected or nay
<img width="931" height="497" alt="image" src="https://github.com/user-attachments/assets/604411d5-3a08-4f80-a339-930e3bfa4ed2" />
4. In the area light documentation, it should be added what to do with light shaders. I couldn't find the information anywhere in-engine. 

5. Attenuation behaves strangely, i don't quite understand. It seems to act behind the light at high value. 
<img width="1376" height="552" alt="image" src="https://github.com/user-attachments/assets/250d5ae5-4053-4827-bef0-b09892849b9a" />

6. as much as i am enjoying toying with negative attenuation, i don't think it should be in the range. We can leave it with `or_less` if we want to, but it shouldn't be presented to the user as a valid "side" of the range
<img width="1487" height="779" alt="image" src="https://github.com/user-attachments/assets/c3d5ffc0-cd72-401e-919a-621b5aee51a3" />

7. Lightmaps seem to bake alright but somebody should check

Next step would be shaders, but i need some help finding the correct formulas again

QbieShay reviewed (approved):
Tested with a testing project and Jamser's bistro. Found some issues, but turned out to be unrelated to area lights. Looks good to me

CookieBadger commented on scene/3d/light_3d.h:
```diff
@@ -86,6 +86,9 @@ class Light3D : public VisualInstance3D {
 	float temperature = 6500.0;
+	Vector2 area_size;
+	Ref<Texture2D> area_texture;
+	bool area_normalize_energy = true;
```
yes.

CookieBadger committed:
Add AreaLight3D

CookieBadger commented:
@clayjohn rebased and addressed your comment.

clayjohn reviewed (approved):
Amazing work! I don't have anything to add after my last few reviews. The current state of this PR is amazing and I consider it ready to merge. 

Let's get this in for 4.7!!

Repiteo commented:
Thanks! Fantastic job!

JekSun97 commented:
> You add it manually (MeshInstance3D, Plane mesh, emissive material, texture). I understand the need for showing it when experimenting, but I wouldn't know how the best way to automate that right now, and in a production 3D environment you would rarely ever want the light plane visible, but rather have some 3D model, like a billboard there instead

I think in another PR we could add a Gizmo as a translucent plane with a direction arrow in the center, this would improve ease of use.
